content/post/2021-01-05-oauth2/index.Rmd

```diff
@@ -0,0 +1,203 @@
+
+Now while this all makes perfect sense when you are developing an actual third-party app, for interactive usage... when all you want is an R package that can be used in scripts running on a server, it's much less clear.
+
+## How do you use OAuth 2.0 in R?
+
+When writing an R package wrapping an API using OAuth 2.0 you'll need the user to grant access to an "app", which will allow to create an access token and a refresh token.
+The access token will then often be passed to the API in a header when making requests, whilst the refresh token would be posted in a query string when the access token needs to be renewed.
+
+Your problem is: how do I imitate a third-party app?
+Thankfully for you, in most cases the complexity can be handled by the httr package.
+For other cases, or if you want to e.g. only use curl, you will have to get creative. :wink:
+
+How does the OAuth 2.0 dance and usage happens with httr?
+
+### OAuth endpoint
+
+* You create an OAuth endpoint (via `httr::oauth_endpoint()`) which is an object holding the URLs to which requests need to be made to request and refresh an access token. Or you use one of the built-in `httr::oauth_endpoints()` to popular APIs.
+
+How do you know what URLs to feed `httr::oauth_endpoint()`? 
+By reading the API docs and looking at examples. :grin:
+E.g. compare [ORCID API docs](https://info.orcid.org/faq/how-does-3-legged-oauth-work/) and [rorcid source code](https://github.com/ropensci/rorcid/blob/afa74daabf18499841e843a755b182c73ab545ea/R/auth.R#L106-L108).
+
+### OAuth app
+
+* You create an OAuth app (via `httr::oauth_app()`) which is an object holding the name, secret and ID of your third-party app. What third-party app?! Say you are using rtweet. You need to register an app in your Twitter developer account e.g. "My Fictional App", you'll get a secret and ID in return. There is no actual app (pfiew!) but Twitter now thinks there is one. The app info will be used by httr when requesting a token, and the user (you) will grant access to their account to "My Fictional App". The server e.g. Twitter will ask you for a callback URL which needs to be `http://localhost:1410/`, that as we'll see later httr will listen to to catch tokens produced by the server.
+
+**In an R package if possible whilst respecting the API rules it is nice to have a built-in OAuth app for quick experiments by the users; and the way for users to specify their own app's information ([Bring Your Own App](https://googledrive.tidyverse.org/articles/articles/bring-your-own-app.html)).**
```

Comment:
Maybe add some commas? 

In an R package, if possible whilst respecting the API rules, it is nice...

Or just remove the "if possible ... API rules" part, I am not actually what that refers to.